### PR TITLE
fix receive msg error

### PR DESCRIPTION
**Pull request by laughxing**
use new header format, handle receive of partial packet.

### CODE REVIEW

File: Source/ChanneldUE/ChanneldConnection.cpp [hunk 1]
```diff
@@ -106,6 +108,25 @@ bool UChanneldConnection::Connect(bool bInitAsClient, const FString& Host, int32
...
+	{
+		UE_LOG(LogChanneld, Error, TEXT("UChanneldConnection::Connect: Set NoDelay failed"));
+	}
+	if(!Socket->SetNonBlocking())
+	{
+		UE_LOG(LogChanneld, Error, TEXT("UChanneldConnection::Connect: Set Non Blocking failed"));
+	}
+	
+	int NewBufferSize = 0;
+	if(!Socket->SetReceiveBufferSize(ReceiveBufferSize, NewBufferSize) || NewBufferSize != ReceiveBufferSize)
```
Comment (indiest) [inline Source/ChanneldUE/ChanneldConnection.cpp:121]: Do we still need the socket's buffer if there's already a receive buffer in ChanneldConnection?

**Comment by indiest** [comment]
@laughxing I'm getting errors of "unmarshalling packet" in channeld when stress testing. Not sure if it's related to this fix. What error did you get?

**Comment by laughxing** [comment]
> @laughxing I'm getting errors of "unmarshalling packet" in channeld when stress testing. Not sure if it's related to this fix. What error did you get?

I also met unmarshalling error when stressing channeld and ds on a linux server,  then I started going through the code and committed this two pr. https://github.com/metaworking/channeld/pull/37 https://github.com/metaworking/channeld-ue-plugin/pull/21, but due to the lack of analysis, I'm not sure if we have the same problem.

File: Source/ChanneldUE/ChanneldConnection.cpp (hunk 1 above)
Comment (indiest) [reply Source/ChanneldUE/ChanneldConnection.cpp:121]: What will happen if we set the socket's buffer size to 0?

**Comment by indiest** [comment]
> > @laughxing I'm getting errors of "unmarshalling packet" in channeld when stress testing. Not sure if it's related to this fix. What error did you get?
> 
> I also met unmarshalling error when stressing channeld and ds on a linux server, then I started going through the code and committed this two pr.[metaworking/channeld#34](https://github.com/metaworking/channeld/pull/34) #21 and this issue [metaworking/channeld#35](https://github.com/metaworking/channeld/issues/35), but due to the lack of analysis, I'm not sure if we have the same problem.

Limiting the number of message in a packet to 1 seems to fix error.

File: Source/ChanneldUE/ChanneldConnection.cpp (hunk 1 above)
Comment (laughxing) [reply Source/ChanneldUE/ChanneldConnection.cpp:121]: kernel need socket buffer to store data before they are read by the application, there's kenerl configs ( /proc/sys/net/core/rmem_default/rmem_max) for it.  according to the [documentation,](https://man7.org/linux/man-pages/man7/socket.7.html) the min value is 256

**Comment by Augkit** [comment]
@laughxing I tested it using a stress tester based on the chat scenario and has solved the packet loss problem!😆

**Comment by laughxing** [comment]
@indiest accedently close this, new pr is submitted https://github.com/metaworking/channeld-ue-plugin/pull/30

**Pull request closed without merging**